In [ ]:
!pip install krippendorff

In [ ]:
import pandas as pd
import numpy as np
import krippendorff
from statsmodels.stats.inter_rater import fleiss_kappa

# Files

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
base_path = '/content/drive/MyDrive/I -TEAM FILES/Datasets/annotation/Stratified Sample/csv/'

In [ ]:
df1 = pd.read_csv(base_path + 'CYN_SS.csv')
df2 = pd.read_csv(base_path + 'ABA_SS.csv')
df3 = pd.read_csv(base_path + 'BRND_SS.csv')

Cleaning

In [ ]:
print("CYN_SS.csv shape:", df1.shape)
print("ABA_SS.csv shape:", df2.shape)
print("BRND_SS.csv shape:", df3.shape)

CYN_SS.csv shape: (27077, 16)
ABA_SS.csv shape: (27077, 16)
BRND_SS.csv shape: (27077, 16)


In [ ]:
print("CYN duplicates:", df1["commentID"].duplicated().sum())
print("ABA duplicates:", df2["commentID"].duplicated().sum())
print("BRND duplicates:", df3["commentID"].duplicated().sum())

CYN duplicates: 300
ABA duplicates: 300
BRND duplicates: 300


In [ ]:
df1 = df1.drop_duplicates(subset="commentID")
df2 = df2.drop_duplicates(subset="commentID")
df3 = df3.drop_duplicates(subset="commentID")

In [ ]:
print("CYN duplicates:", df1["commentID"].duplicated().sum())
print("ABA duplicates:", df2["commentID"].duplicated().sum())
print("BRND duplicates:", df3["commentID"].duplicated().sum())

CYN duplicates: 0
ABA duplicates: 0
BRND duplicates: 0


In [ ]:
print("CYN_SS.csv shape:", df1.shape)
print("ABA_SS.csv shape:", df2.shape)
print("BRND_SS.csv shape:", df3.shape)

CYN_SS.csv shape: (26777, 16)
ABA_SS.csv shape: (26777, 16)
BRND_SS.csv shape: (26777, 16)


In [ ]:
df1 = df1.rename(columns={"label": "lbl1"})
df2 = df2.rename(columns={"label": "lbl2"})
df3 = df3.rename(columns={"label": "lbl3"})

In [ ]:
merged_df = df1.drop(columns=["lbl1"]).merge(
    df1[["commentID", "lbl1"]], on="commentID"
).merge(
    df2[["commentID", "lbl2"]], on="commentID"
).merge(
    df3[["commentID", "lbl3"]], on="commentID"
)

In [ ]:
print("Merged DataFrame shape:", merged_df.shape)

Merged DataFrame shape: (26777, 18)


In [ ]:
merged_df.to_csv("merged_annotations.csv", index=False)

# IRR

In [ ]:
merged_df = pd.read_csv("/content/merged_annotations.csv")

In [ ]:
merged_df

,userID,facebookPost,newsOutlet,postDate,postTime,commentID,commentText,commentDate,commentTime,attachment,isReply,parentCommentID,depth,likes,features,lbl1,lbl2,lbl3
0,1021844,"“'Pag walang korapsyon, lahat ng pera ng taong...",gma,2022-04-30,22:50:00,Y29tbWVudDoxMDE2MDkzMTkyNDk0Njk3N18yODIwOTUwMz...,Asa ka naman. Dilawan! Campaign manager mo nga...,2022-04-30,23:54:14,NaN,False,NaN,0,40,NaN,Troll,Troll,Human
1,1021844,"“'Pag walang korapsyon, lahat ng pera ng taong...",gma,2022-04-30,22:50:00,Y29tbWVudDoxMDE2MDkzMTkyNDk0Njk3N18xMjI0NjM4OT...,punta k po sa supreme court bukas. Latag mo n ...,2022-04-30,23:56:41,NaN,True,Y29tbWVudDoxMDE2MDkzMTkyNDk0Njk3N18yODIwOTUwMz...,2,0,NaN,Troll,Troll,Human
2,1021844,"“'Pag walang korapsyon, lahat ng pera ng taong...",gma,2022-04-30,22:50:00,Y29tbWVudDoxMDE2MDkzMTkyNDk0Njk3N183MjE5ODA5OD...,"guard, may lutang dito. 😝",2022-04-30,23:57:40,NaN,True,Y29tbWVudDoxMDE2MDkzMTkyNDk0Njk3N18yODIwOTUwMz...,2,1,NaN,Troll,Troll,Human
3,1021844,"“'Pag walang korapsyon, lahat ng pera ng taong...",gma,2022-04-30,22:50:00,Y29tbWVudDoxMDE2MDkzMTkyNDk0Njk3N18zODM1MTk1ND...,tablea scripted😝😝,2022-04-30,23:58:49,NaN,True,Y29tbWVudDoxMDE2MDkzMTkyNDk0Njk3N18yODIwOTUwMz...,2,0,NaN,Troll,Troll,Human
4,1021844,"“'Pag walang korapsyon, lahat ng pera ng taong...",gma,2022-04-30,22:50:00,Y29tbWVudDoxMDE2MDkzMTkyNDk0Njk3N181MTM4MzkyMj...,"Ask ko lang, bakit nsa Tarlac Rally si Kris? P...",2022-05-01,0:00:57,NaN,True,Y29tbWVudDoxMDE2MDkzMTkyNDk0Njk3N18yODIwOTUwMz...,2,0,NaN,Troll,Troll,Human
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
26772,9987708,Nabunyag sa isang pag-aaral na maraming tagasu...,abscbn,2022-04-26,22:30:00,Y29tbWVudDoxMDE2MDI3MzI2ODgxNTE2OF8yMzk1NDUxMD...,Hanggana ngayon ganon parin ang Abiascbn,2022-04-28,18:56:32,NaN,False,NaN,0,0,NaN,Human,Human,Human
26773,9994968,#FromTheArchives: 'I do not know how the syste...,Rappler,2022-04-28,21:01:00,Y29tbWVudDozNjcyMDU2MjU0NDE1MTJfMzY3MjcwNzg1ND...,#FreeLeilaNow,2022-04-28,21:47:41,NaN,False,NaN,0,0,NaN,Human,Cyborg,Human
26774,9994968,"Eight days before the May 9 elections, former ...",Inquirer.net,2022-05-01,21:43:00,Y29tbWVudDozMDUxMTk0MDg0NjU2ODhfMzA1MTI2MjkxNz...,#10RobredoForPresident #7PangilinanForVicePres...,2022-05-01,21:56:11,Photo,False,NaN,0,28,NaN,Human,Cyborg,Human
26775,9994968,FREE LEILA NOW!' LOOK: Senatorial bets Neri C...,Inquirer.net,2022-04-28,22:01:00,Y29tbWVudDozMDI4MTEwNTg2OTY1MjNfMzAyODEyOTg1Mz...,#FreeLeilaNow,2022-04-28,22:04:43,NaN,False,NaN,0,6,NaN,Human,Cyborg,Human


In [ ]:
# Extract label columns
annotations = merged_df[["lbl1", "lbl2", "lbl3"]]

In [ ]:
# Encode string labels into integers
label_mapping = {label: idx for idx, label in enumerate(pd.unique(annotations.values.ravel()))}
encoded_annotations = annotations.replace(label_mapping)

<ipython-input-63-2c24807c9b9b>:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  encoded_annotations = annotations.replace(label_mapping)


Krippendorff’s Alpha

In [ ]:
# Transpose to shape: raters x items
kripp_alpha_input = encoded_annotations.T.values

In [ ]:
# Compute Krippendorff's Alpha (nominal level)
kripp_alpha = krippendorff.alpha(reliability_data=kripp_alpha_input, level_of_measurement='nominal')
print("Krippendorff's Alpha:", kripp_alpha)

Krippendorff's Alpha: 0.3710660513637206


Fleiss’ Kappa

In [ ]:
# Count number of times each category was chosen per item
fleiss_input = encoded_annotations.apply(lambda row: row.value_counts(), axis=1).fillna(0)

In [ ]:
# Ensure consistent order of columns (i.e., categories)
for col in range(len(label_mapping)):
    if col not in fleiss_input.columns:
        fleiss_input[col] = 0
fleiss_input = fleiss_input[sorted(fleiss_input.columns)]

In [ ]:
# Convert to matrix
fleiss_matrix = fleiss_input.values.astype(int)

In [ ]:
# Compute Fleiss' Kappa
fleiss_score = fleiss_kappa(fleiss_matrix)
print("Fleiss' Kappa:", fleiss_score)

Fleiss' Kappa: 0.37105822198554755


**Krippendorff's Alpha:** *0.371*

**Fleiss' Kappa:** *0.371*

3 annotators had a fair agreement

# Final Label

In [ ]:
# Function to compute majority vote for each row
def majority_vote(row):
    return row.mode().iloc[0]  # picks the most common label

In [ ]:
# Assign final label
merged_df["fin_lbl"] = merged_df[["lbl1", "lbl2", "lbl3"]].apply(majority_vote, axis=1)

In [ ]:
print(merged_df[["lbl1", "lbl2", "lbl3", "fin_lbl"]].head())

    lbl1   lbl2   lbl3 fin_lbl
0  Troll  Troll  Human   Troll
1  Troll  Troll  Human   Troll
2  Troll  Troll  Human   Troll
3  Troll  Troll  Human   Troll
4  Troll  Troll  Human   Troll


In [ ]:
merged_df.to_csv("sampled_data(fin_lbl).csv", index=False)

# Erronous Labels (pls attend to asap)

In [ ]:
df = pd.read_csv(base_path + 'sampled_data(fin_lbl).csv')

In [ ]:
df_no_duplicates = df.drop_duplicates(subset=['userID', 'lbl1', 'lbl2', 'lbl3', 'fin_lbl'])

In [ ]:
len(df_no_duplicates)

880

In [ ]:
check_df = df_no_duplicates[['userID', 'lbl1', 'lbl2', 'lbl3', 'fin_lbl']]
check_df

,userID,lbl1,lbl2,lbl3,fin_lbl
0,1021844,Troll,Troll,Human,Troll
17,1032253,Human,Human,Human,Human
28,1039664,Cyborg,Troll,Cyborg,Cyborg
147,1050970,Human,Troll,Human,Human
149,1052054,Human,Troll,Human,Human
...,...,...,...,...,...
26666,9967534,Human,Human,Human,Human
26671,9972201,Human,Troll,Troll,Troll
26757,9985661,Human,Troll,Human,Human
26772,9987708,Human,Human,Human,Human


In [ ]:
duplicated_entries = check_df[check_df.duplicated(subset=['userID'], keep=False)]

duplicated_entries

,userID,lbl1,lbl2,lbl3,fin_lbl
163,1069688,Troll,Cyborg,Troll,Troll
190,1069688,Troll,Cyborg,Human,Cyborg
828,1287976,Troll,Spam Bot,Spam Bot,Spam Bot
900,1287976,Troll,Spam Bot,Human,Human
901,1310539,Human,Troll,Human,Human
...,...,...,...,...,...
26229,9770805,Human,Cyborg,Spam Bot,Cyborg
26426,9866970,Human,Troll,Human,Human
26502,9866970,Human,Troll,Troll,Troll
26598,9930324,Cyborg,Cyborg,Cyborg,Cyborg


In [ ]:
duplicated_count = duplicated_entries['userID'].value_counts()

print(duplicated_count)

userID
5878188    3
5825891    3
6357694    3
9730831    3
1310539    2
1069688    2
1287976    2
3985325    2
3270539    2
3580452    2
1413537    2
5527226    2
4528697    2
4036152    2
5668974    2
5727526    2
5689120    2
5679836    2
3608019    2
5727983    2
6631276    2
7963957    2
6844164    2
8413149    2
9434682    2
9480450    2
8339864    2
9550229    2
9651361    2
9699054    2
9686729    2
9730744    2
9770805    2
9866970    2
9930324    2
Name: count, dtype: int64


Merge votes from all duplicates, then recompute majority vote per userID

In [ ]:
# Group by userID and gather all annotations
grouped = duplicated_entries.groupby('userID').agg({
    'lbl1': lambda x: list(x),
    'lbl2': lambda x: list(x),
    'lbl3': lambda x: list(x),
})

In [ ]:
# Flatten the labels into a single list per userID
grouped['all_labels'] = grouped.apply(lambda row: row['lbl1'] + row['lbl2'] + row['lbl3'], axis=1)

In [ ]:
# Majority vote again
def true_majority(labels):
    return pd.Series(labels).mode().iloc[0]  # Most frequent label

In [ ]:
grouped['recomputed_final_label'] = grouped['all_labels'].apply(true_majority)

In [ ]:
print(grouped[['all_labels', 'recomputed_final_label']])

                                                all_labels  \
userID                                                       
1069688       [Troll, Troll, Cyborg, Cyborg, Troll, Human]   
1287976  [Troll, Troll, Spam Bot, Spam Bot, Spam Bot, H...   
1310539         [Human, Human, Troll, Troll, Human, Troll]   
1413537         [Troll, Troll, Human, Human, Troll, Human]   
3270539         [Troll, Troll, Troll, Troll, Troll, Human]   
3580452        [Human, Human, Troll, Troll, Human, Cyborg]   
3608019         [Human, Human, Human, Human, Human, Troll]   
3985325         [Human, Human, Human, Human, Troll, Human]   
4036152        [Human, Human, Troll, Troll, Troll, Cyborg]   
4528697      [Troll, Troll, Troll, Troll, Spam Bot, Human]   
5527226  [Spam bot, Spam bot, Troll, Cyborg, Troll, Cyb...   
5668974       [Human, Human, Human, Cyborg, Human, Cyborg]   
5679836     [Cyborg, Cyborg, Cyborg, Human, Cyborg, Human]   
5689120         [Human, Human, Human, Troll, Human, Human]   
5727526 

In [ ]:
duplicated_check = grouped.index.duplicated().sum()

print(f"Number of duplicated userIDs after recomputing: {duplicated_check}")

Number of duplicated userIDs after recomputing: 0


# Removing Erronous Labels

In [ ]:
df_finLbl = pd.read_csv(base_path + 'sampled_data(fin_lbl).csv')

In [ ]:
grouped = grouped.reset_index()

In [ ]:
df_fixed = df.merge(grouped[['userID', 'recomputed_final_label']], on='userID', how='left')

In [ ]:
df_fixed['fin_lbl'] = df_fixed['recomputed_final_label'].combine_first(df_fixed['fin_lbl'])

In [ ]:
df_fixed = df_fixed.drop(columns=['recomputed_final_label'])

In [ ]:
df_fixed = df_fixed.drop_duplicates(subset=['userID'])

In [ ]:
print("Final shape after fixing:", df_fixed.shape)
print("Number of unique userIDs:", df_fixed['userID'].nunique())

Final shape after fixing: (841, 19)
Number of unique userIDs: 841


In [ ]:
df_fixed.to_csv("sampled_data(final_lbl).csv", index=False)

In [ ]:
import pandas as pd
df = pd.read_csv("sampled_data_fixed(final_lbl).csv")

# Standardize final label names (e.g., unify "Spam bot" and "Spam Bot")
df['fin_lbl_clean'] = df['fin_lbl'].str.strip().str.title()

# Compute value counts and percentages
final_counts = df['fin_lbl_clean'].value_counts()
final_percentages = (final_counts / final_counts.sum()) * 100

# Combine into one DataFrame
label_distribution = pd.DataFrame({
    'Count': final_counts,
    'Percentage': final_percentages.round(2)
})

# Print or inspect the result
print(label_distribution)

               Count  Percentage
fin_lbl_clean                   
Human            453       53.86
Troll            206       24.49
Cyborg           156       18.55
Spam Bot          26        3.09
